### DISCUSSION TITLE
I cannot build Thorium!

### DISCUSSION THREAD

**Discussion by ghost** [Alex313031/thorium#172] [Q&A] [answered]
Env:MacOS13.2.1,XcodeSDK14.3.1

I ran the following command to do the build
`gn args out/thorium`
However, GN output the following error
`
ERROR at //media/video/BUILD.gn:69:7: Undefined identifier
  if (enable_platform_vvc) {
      ^------------------
See //media/media_options.gni:351:3: which caused the file to be included.
  "//media/video",
  ^--------------
`
I looked at the file where the problem supposedly occurred and did not see the problem.
I am following "https://github.com/Alex313031/Thorium/blob/main/docs/BUILDING.md" for the build.
I'll post a photo that might be relevant for reference.

<img width="468" alt="スクリーンショット 2023-06-20 8 17 42" src="https://github.com/Alex313031/thorium/assets/121844712/f5a86127-b3f2-4941-992d-caa2f8da39ba">
<img width="338" alt="スクリーンショット 2023-06-20 8 19 24" src="https://github.com/Alex313031/thorium/assets/121844712/e3ac48d7-c417-421f-b049-546820b3fff7">
<img width="568" alt="スクリーンショット 2023-06-20 8 20 17" src="https://github.com/Alex313031/thorium/assets/121844712/c3d125ac-de92-485a-8e1d-bf632c04b9d8">
<img width="493" alt="スクリーンショット 2023-06-20 8 20 36" src="https://github.com/Alex313031/thorium/assets/121844712/d0026f85-36f1-4951-ac7c-30d87dec87e6">

**Comment by gz83** [comment]
@barappe 

You cannot compile Thorium with the latest commit in the current repo, because the repo has not yet been rebased.

If you need to compile Thorium, you need to roll back the commit to >[a49c6332f3ae7cdb1567248a0c14ee0b44929ac3](https://github.com/Alex313031/thorium/commit/a49c6332f3ae7cdb1567248a0c14ee0b44929ac3)

Then compile according to the guidelines in the repo and Chromium's official documentation.

**Reply by ghost** [reply]
I rolled back the commit as you instructed, but I get the same error.

Let me know if you need any other log files, etc.

**Reply by gz83** [reply]
Please run `git log` in the `chromium/src` directory and send a screenshot

In addition, you may need to undo the changes made in the chromium/src directory

@barappe

**Reply by ghost** [reply]
Here is the result of the run
`
commit 64dcbbd5e7191b268a569f7acae7dc4ccc8cceb0 (HEAD, origin/main)
Author: kylechar <kylechar@chromium.org>
Date:   Tue Jun 20 14:34:09 2023 +0000

    Fix AllowBypassRenderPassQuads clip
    
    The intersection of bypass quads clip_rect was done in the wrong
    coordinate space when there was a chain of bypassed render pass and
    multiple intermediate render pass coordinate spaces were merged
    together. Always transform back to the intermediate coordinate space,
    the same coordinate space as the first bypassed render pass, to fix it.
    
    Bug: 1455497
    Change-Id: Iae6ddacadada1404260747808757286b9f5d319a
    Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4627213
    Reviewed-by: Michael Ludwig <michaelludwig@google.com>
    Commit-Queue: Kyle Charbonneau <kylechar@chromium.org>
    Cr-Commit-Position: refs/heads/main@{#1160013}

commit 92d362a1338692edb098c5f1451808c6de307c31
Author: chromium-autoroll <chromium-autoroll@skia-public.iam.gserviceaccount.com>
Date:   Tue Jun 20 14:33:19 2023 +0000

    Roll Chrome Win64 PGO Profile
    
    Roll Chrome Win64 PGO profile from chrome-win64-main-1687251359-ae83a8f0f625c6c4211579dea518a547f314ad8c.profdata to chrome-win64-main-1687262198-e7806ce3dab7501254112f074f2516eb8dc6ad92.profdata
    
    If this roll has caused a breakage, revert this CL and stop the roller
    using the controls here:
    https://autoroll.skia.org/r/pgo-win64-chromium
    Please CC chrome-brapp-engprod@google.com,pgo-profile-sheriffs@google.com on the revert to ensure that a human
    is aware of the problem.
    
    To file a bug in Chromium main branch: https://bugs.chromium.org/p/chromium/issues/entry
    
    To report a problem with the AutoRoller itself, please file a bug:
    https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug
    
    Documentation for the AutoRoller is here:
    https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md
    
    Cq-Include-Trybots: luci.chrome.try:win64-chrome
    Tbr: pgo-profile-sheriffs@google.com
    Change-Id: I2745498e17ae97a5e54881d0bc6c6e897b066a24
    Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4628168
    Bot-Commit: chromium-autoroll <chromium-autoroll@skia-public.iam.gserviceaccount.com>
    Commit-Queue: chromium-autoroll <chromium-autoroll@skia-public.iam.gserviceaccount.com>
    Cr-Commit-Position: refs/heads/main@{#1160012}

commit 91fade9e25fa009b047d2f69deee112313648c16
Author: Veronique Nguyen <veronguyen@google.com>
Date:   Tue Jun 20 14:31:49 2023 +0000

    [iOS] Fix Password Bottom Sheet UI issues
    
    - Credential cell cropped at the bottom
    - Fix Password Details view title view sizing issue
    
    Bug: 1451311, 1446351
    Change-Id: I8c21eb6c56d7992f462f8ad50a431cff595bfd0a
    Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/4608837
    Reviewed-by: Sylvain Defresne <sdefresne@chromium.org>
    Commit-Queue: Veronique Nguyen <veronguyen@google.com>
    Cr-Commit-Position: refs/heads/main@{#1160011}

commit d88eb124630e333d40352f232ee1b5f2f7596abc
Author: Victor Hugo Vianna Silva <victorvianna@google.com>
Date:   Tue Jun 20 14:25:12 2023 +0000

    [uno] Integrate NTP identity disc with sign-in sheet
    
    Behavior changes are behind flag. Have the identity disc lead to the
    sign-in sheet instead of the sync consent dialog. The case when there

`
Do I need to get more commit history?

**Comment by Alex313031** [comment]
@barappe @gz83  What you should do is checkout thorium at `a49c6332f3ae7cdb1567248a0c14ee0b44929ac3`, then make sure that the thorium-libjxl submodule is checked out to the proper revision with `git submodule update`.

The instructions need to be updated, and I also need to make a new .md file specifically for Mac. Sorry. Until then:

What you would do then, is run `./VERSION.sh`, then `./setup.sh --mac`, and finally, run your `gn args out/thorium` making sure to use the correct PGO profile at the bottom (which should have been downloaded with the VERSION script). If you want Google Sync support you will need to add the API Keys at the top of your args. I have intentionally excluded these from the thorium repo so that Google doesn't come after me. To get them, you can run the `./fetch_api_keys.sh` script that is in the //infra directory. This simply uses wget to download an API_KEYS.txt file from a separate GitHub account and repo that I made for this purpose.

For most of these scripts to work, you will need both thorium and chromium in your home dir. For example, on Linux and Mac they should be ~/thorium and ~/chromium.

Keep in mind that the args for ARM and x64 MacOS are seperate. They are in //other/Mac/mac_args.gn and //other/Mac/mac_ARM_args.gn.

**Reply by ghost** [reply]
After following the instructions, the error changed as follows
![image](https://github.com/Alex313031/thorium/assets/121844712/f54c5b72-a8f5-43a4-b881-13d84e66912e)

I have attached the args.gn file for your reference.
![image](https://github.com/Alex313031/thorium/assets/121844712/b88376c7-ced4-4df2-b569-196a1156b43a)

Thanks for the advice.

**Reply by Alex313031** [reply]
@midzer Can you reproduce this on your Mac? It seems he is doing everything right but still getting errors.

@barappe I do not have a Mac, so I cant test for myself unlike the other platforms (which is frustrating for bug triage) One thing you might test, to see if any args are causing this, is try to make a build with a minimal args.gn. Something like:
```
target_os = "mac"
target_cpu = "x64"
v8_target_cpu = "x64"
use_system_xcode = true
is_official_build = true
proprietary_codecs = true
ffmpeg_branding = "Chrome"
```
Changing `target_cpu` and `v8_target_cpu` to "arm64" if building for M1 Mac.

**Reply by ghost** [reply]
I followed the advice and the results changed just a little bit.
![image](https://github.com/Alex313031/thorium/assets/121844712/2b2ac5f0-267c-40e7-8fd9-7b3dae60c34d)

**Reply by midzer** [reply]
Hi @barappe. First, try a plain Chromium build. If this works, add a minimal args.gn as @Alex313031 already suggested. If this fails, I'm gonna double check our suggested `args.gn` for Mac ARM64.

**Reply by gz83** [reply]
I think you need to re-follow the repo and the guidelines provided by the official documentation, re-fetch the source code of the corresponding version, and make sure the thorium repository is up-to-date, and then try to compile according to the opinions of other team members

@barappe

**Reply by ghost** [reply]
@midzer I had success with the normal Chromium build a month ago, but will try again.
@gz83 I do check the Xcode SDK version and Thromium version, and I re-fetch Chromium (I use --nohistory option) every time.

**Reply by gz83** [reply, marked as answer]
I think you should try to follow these steps:

1. Make sure your device meets the requirements listed in the official chromium documentation -> [https://chromium.googlesource.com/chromium/src/+/main/docs/mac_build_instructions.md#system- requirements](url)
2. Install depot_tools -> [https://chromium.googlesource.com/chromium/src/+/main/docs/mac_build_instructions.md#install](url)
3. Get the full chromium source code (including history and hooks)->[https://chromium.googlesource.com/chromium/src/+/main/docs/mac_build_instructions.md#get-the-code](url)
4. Re-cloning of thorium repo->`git clone https://github.com/Alex313031/thorium.git --recurse-submodules`
5. Execute `./VERSION.sh` to switch the chromium source code to the branch currently used by thorium, and then use `./setup.sh --mac` to copy the files needed by thorium to the chromium source directory
6. Choose the corresponding args.gn file (//other/Mac/mac_args.gn or //other/Mac/mac_ARM_args.gn) according to your device's architecture and copy the contents of the file in advance
7. Execute `gn args out/thorium` under the chromium source code directory (chromium/src), paste the contents of the args.gn file you just copied in the editor that opens, then save and exit the editor, and wait for gn to generate the various files(target files, intermediate files, etc.)
8. If you need to use the synchronization feature, you will also need to pass `./fetch_api_keys.sh` to get the API and KEY and place them in the args.gn file
9. Use `build_mac.sh` to compile thorium, and after the compilation is done, you can use `build_dmg.sh` to compile dmg

Please note that the above are only some general steps, the specific operation needs to be determined according to the actual situation.

@barappe

**Reply by ghost** [reply]
As a result, the build was successful. The reason for the lack of success seems to be due to the use of the --nohistory option; I missed the fact that Checkout was not working.
Thanks for your help!
